prover/src/circuits/tests.rs

```diff
@@ -1,68 +1,98 @@
+        MultilinearAsUnivariate,
+    },
 };
 const K: u32 = 17;
 const N_CURRENCIES: usize = 2;
 const N_USERS: usize = 1 << 16;
 
-pub fn run_plonkish_backend<F, Pb, T, C>(
-    num_vars: usize,
-    circuit_fn: impl Fn(usize) -> (PlonkishCircuitInfo<F>, C),
-) where
-    F: PrimeField + Hash + Serialize + DeserializeOwned,
-    Pb: PlonkishBackend<F>,
-    T: TranscriptRead<<Pb::Pcs as PolynomialCommitmentScheme<F>>::CommitmentChunk, F>
-        + TranscriptWrite<<Pb::Pcs as PolynomialCommitmentScheme<F>>::CommitmentChunk, F>
-        + InMemoryTranscript<Param = ()>,
-    C: PlonkishCircuit<F>,
-{
-    let (circuit_info, circuit) = circuit_fn(num_vars);
+pub fn seeded_std_rng() -> impl RngCore + CryptoRng {
+    StdRng::seed_from_u64(OsRng.next_u64())
+}
+
+#[test]
+fn test_summa_hyperplonk() {
+    type ProvingBackend = HyperPlonk<MultilinearKzg<Bn256>>;
+    let entries = generate_dummy_entries::<N_USERS, N_CURRENCIES>().unwrap();
+    let circuit = SummaHyperplonk::<N_USERS, N_CURRENCIES>::init(entries.to_vec());
+    let num_vars = K;
+
+    let circuit_fn = |num_vars| {
+        let circuit = Halo2Circuit::<Fp, SummaHyperplonk<N_USERS, N_CURRENCIES>>::new::<
+            ProvingBackend,
+        >(num_vars, circuit.clone());
+        (circuit.circuit_info().unwrap(), circuit)
+    };
+
+    let (circuit_info, circuit) = circuit_fn(num_vars as usize);
     let instances = circuit.instances();
 
-    let param = Pb::setup(&circuit_info, seeded_std_rng()).unwrap();
+    let param = ProvingBackend::setup(&circuit_info, seeded_std_rng()).unwrap();
+
+    let (prover_parameters, verifier_parameters) =
+        ProvingBackend::preprocess(&param, &circuit_info).unwrap();
 
-    let (pp, vp) = Pb::preprocess(&param, &circuit_info).unwrap();
+    let (witness_polys, proof_transcript) = {
+        let mut proof_transcript = Keccak256Transcript::new(());
 
-    let proof = {
-        let mut transcript = T::new(());
-        Pb::prove(&pp, &circuit, &mut transcript, seeded_std_rng()).unwrap();
-        transcript.into_proof()
+        let witness_polys = ProvingBackend::prove(
+            &prover_parameters,
+            &circuit,
+            &mut proof_transcript,
+            seeded_std_rng(),
+        )
+        .unwrap();
+        (witness_polys, proof_transcript)
     };
 
-    let result = {
-        let mut transcript = T::from_proof((), proof.as_slice());
-        Pb::verify(&vp, instances, &mut transcript, seeded_std_rng())
+    let num_points = 3;
+
+    let proof = proof_transcript.into_proof();
+
+    let mut transcript;
+    let result: Result<(), plonkish_backend::Error> = {
+        transcript = Keccak256Transcript::from_proof((), proof.as_slice());
+        ProvingBackend::verify(
+            &verifier_parameters,
+            instances,
+            &mut transcript,
+            seeded_std_rng(),
+        )
     };
     assert_eq!(result, Ok(()));
 
     let wrong_instances = instances[0]
```

Comment:
Sure!

Reply:
I suppose there's some misunderstanding, instances[0] is not the first value of the instances but the whole instance vector. It is because `circuit.instances()` returns a slice of vectors as required by the `PlonkishCircuit` trait. So here every instance value gets modified to a wrong one, not just the first one.

Reply:
> I suppose there's some misunderstanding, instances[0] is not the first value of the instances but the whole instance vector. It is because `circuit.instances()` returns a slice of vectors as required by the `PlonkishCircuit` trait. So here every instance value gets modified to a wrong one, not just the first one.

Oh, I see. The iteration increments all instances by 1: [0 + 1, total_balance_1 + 1, total_balance_2 + 1]. If this is correct, Could we add the `skip` method to the iteration to modify only the total balances, like this: [0, total_balance_1 + 1, total_balance_2 + 1]? This change would help us distinguish whether errors originate from the username or the total balances.